File: templates/hooks/on-error.py
#!/usr/bin/env python3
"""on-error.py — Stuck Thread (V2)

Hook: errorOccurred (Copilot CLI)
Writes a thread with status="stuck" and captures the error message.
Appends JSONL audit line.
"""

import json
import os
import sys
import uuid
from datetime import datetime, timezone


def _safe_exit():
    """Output valid JSON and exit cleanly — used when imports fail."""
    try:
        print(json.dumps({"output": ""}))
    except Exception:
        print('{"output": ""}')
    sys.exit(0)


try:
    from sticky_utils import get_memory_path, load_json, save_json, append_audit_line, get_user, detect_tool, get_session_id
except Exception:
    if __name__ == "__main__":
        _safe_exit()


# ── Main───────────────────────────────────────────────────

def main():
    try:
        hook_input = json.loads(sys.stdin.read()) if not sys.stdin.isatty() else {}
    except Exception:
        hook_input = {}

    session_id = get_session_id(hook_input)
    tool_name = hook_input.get("tool_name", os.environ.get("TOOL_NAME", "unknown"))
    if tool_name == "unknown":
        tool_name = detect_tool(hook_input)
    error_msg = hook_input.get("error", hook_input.get("message", "Unknown error"))[:200]
    user = get_user()
    now = datetime.now(timezone.utc).isoformat()

    memory_path = get_memory_path()
    memory = load_json(memory_path, {"version": "2", "project": "", "threads": []})

    # Find existing threadfor this session or create a new one
    threads = memory.setdefault("threads", [])
    existing = None
    for thread in threads:
        if thread.get("session_id") == session_id:
            existing = thread
            break

    if existing:
        existing["status"] = "stuck"
        existing["last_note"] = error_msg
        existing["last_activity_at"] = now
        # Track as a failed approach
        failed = existing.setdefault("failed_approaches", [])
        failed.append({"description": error_msg[:150], "error": error_msg[:100]})
        failed[:] = failed[-5:]  # cap at 5
    else:
        thread = {
            "id": str(uuid.uuid4()),
            "user": user,
            "project": memory.get("project", ""),
            "status": "stuck",
            "branch": "",
            "created_at": now,
            "closed_at": None,
            "last_activity_at": now,
            "files_touched": [],
            "last_note": error_msg,
            "narrative": "",
            "failed_approaches": [{"description": error_msg[:150], "error": error_msg[:100]}],
            "handoff_summary": "",
            "related_session_ids": [],
            "tool": tool_name,
            "session_id": session_id,
        }
        threads.append(thread)

    # Append JSONL audit line
    append_audit_line({
        "type": "error",
        "user": user,
        "ts": now,
        "session_id": session_id,
        "error": error_msg,
        "tool": tool_name,
    })

    save_json(memory_path, memory)
    print(json.dumps({"output": ""}))


if __name__ == "__main__":
    try:
        main()
    except BaseException:
        _safe_exit()

File: templates/hooks/on-error.js
#!/usr/bin/env node
"use strict";
/**
 * on-error.js — Stuck Thread (V2)
 *
 * Hook: errorOccurred (Copilot CLI) / PostToolUseFailure (Claude Code)
 * Writes a thread with status="stuck" and captures the error message.
 * Appends JSONL audit line.
 */

const crypto = require("crypto");

function _safeExit() {
  try {
    process.stdout.write(JSON.stringify({ output: "" }) + "\n");
  } catch (_) {
    process.stdout.write('{"output":""}\n');
  }
  process.exit(0);
}

let getMemoryPath, loadJson, saveJson, appendAuditLine, getUser, detectTool, getSessionId;
try {
  ({
    getMemoryPath,
    loadJson,
    saveJson,
    appendAuditLine,
    getUser,
    detectTool,
    getSessionId,
  } = require("./sticky-utils.js"));
} catch (_) {
  _safeExit();
}

function main() {
  let hookInput = {};
  try {
    if (!process.stdin.isTTY) {
      const raw = require("fs").readFileSync(0, "utf-8").trim();
      if (raw) {
        hookInput = JSON.parse(raw);
      }
    }
  } catch (_) {
    hookInput = {};
  }

  const sessionId = getSessionId(hookInput);
  let toolName = hookInput.tool_name || process.env.TOOL_NAME || "unknown";
  if (toolName === "unknown") {
    toolName = detectTool(hookInput);
  }
  const errorMsg = (hookInput.error || hookInput.message || "Unknown error").substring(0, 200);
  const user = getUser();
  const now = new Date().toISOString();

  const memoryPath = getMemoryPath();
  const memory = loadJson(memoryPath, { version: "2", project: "", threads: [] });

  if (!Array.isArray(memory.threads)) {
    memory.threads = [];
  }
  const threads = memory.threads;

  let existing = null;
  for (const thread of threads) {
    if (thread.session_id === sessionId) {
      existing = thread;
      break;
    }
  }

  if (existing) {
    existing.status = "stuck";
    existing.last_note = errorMsg;
    existing.last_activity_at = now;
    if (!Array.isArray(existing.failed_approaches)) {
      existing.failed_approaches = [];
    }
    existing.failed_approaches.push({
      description: errorMsg.substring(0, 150),
      error: errorMsg.substring(0, 100),
    });
    // Cap at 5 entries
    if (existing.failed_approaches.length > 5) {
      existing.failed_approaches = existing.failed_approaches.slice(-5);
    }
  } else {
    const thread = {
      id: crypto.randomUUID(),
      user: user,
      project: memory.project || "",
      status: "stuck",
      branch: "",
      created_at: now,
      closed_at: null,
      last_activity_at: now,
      files_touched: [],
      last_note: errorMsg,
      narrative: "",
      failed_approaches: [{ description: errorMsg.substring(0, 150), error: errorMsg.substring(0, 100) }],
      handoff_summary: "",
      related_session_ids: [],
      tool: toolName,
      session_id: sessionId,
    };
    threads.push(thread);
  }

  appendAuditLine({
    type: "error",
    user: user,
    ts: now,
    session_id: sessionId,
    error: errorMsg,
    tool: toolName,
  });

  saveJson(memoryPath, memory);
  const statusMsg = `[STICKY-NOTE] Marked thread as STUCK - ${errorMsg.substring(0, 80)}`;
  try {
    process.stdout.write(JSON.stringify({ output: statusMsg }) + "\n");
  } catch (_) {
    process.stdout.write('{"output":""}\n');
  }
  process.exit(0);
}

try {
  main();
} catch (_) {
  _safeExit();
}
